docs/rfcs/749-unify-path-behavior.md

```diff
@@ -0,0 +1,151 @@
+- Author: JinnyYi <github.com/JinnyYi>
+- Start Date: 2021-09-08
+- RFC PR: [beyondstorage/go-storage#749](https://github.com/beyondstorage/go-storage/pull/749)
+- Tracking Issue: [beyondstorage/go-storage#0](https://github.com/beyondstorage/go-storage/issues/0)
+
+# GSP-749: Unify Path Behavior
+
+Previous Discussion:
+
+- [Specify the behavior of Path](https://forum.beyondstorage.io/t/topic/195)
+- [OS-specific path separator problem](https://github.com/beyondstorage/go-service-ftp/issues/22)
+
+## Background
+
+For file system, paths are used extensively to represent the directory/file relationships. Resources can be represented by either absolute or relative paths. And the directory separator is platform specific.
+
+For object storage, the key of an object uniquely identifies the object in a bucket. Key name prefixes and delimiters could be used to group objects or simulate directory.
+
+In our services, path could be a file path for file system, or a file hosting service, also it could be an object for object storage. So we need to unify path behavior for different platforms and storages.
+
+## Proposal
+
+### Absolute path and relative path
+
+All our services should support two kinds of path:
+
+- Absolute path: must include the root directory.
+- Relative path: a relative path based on the working directory.
+
+### Directory separator
+
+From user side:
+
+- System-related directory separator including `\` for Unix and `//` for Windows SHOULD be allowed.
+
+From go-storage side:
+
+- go-storage SHOULD be able to tolerate mixed use of directory separator and replace each separator with a slash (`/`) character could be generated for the input path.
+
+From service side:
+
+- Services SHOULD replace separator in the passed-in path to the current system-related directory separator at the beginning of operations.
+  
+### Implementation
+
+**work_dir**
+
+The pair `work_dir` is used to specific the working directory for service or storage. For object storage, it's the simulated directory or prefix of the object key.
+
+- `work_dir` SHOULD be an absolute path.
+- `work_dir` SHOULD be default to `/` if not set.
+- `work_dir` MUST be Unix style path for object storage services.
+
+Take the example of setting working directory: 
+
+```go
+// set work_dir for init
+pairs.WithWorkDir("/path/to/workdir")
+
+// set default value for `work_dir`
+store := &Storage{
+	// ... 
+	workDir: "/",
+}
+if opt.HasWorkDir {
+	store.workDir = opt.WorkDir
+}
+```
+
+**path**
+
+The field `path` is used as an input parameter for operation, indicating the file path for file system, object key for object storage or prefix filter for `List` operator.
+
+- `path` could be absolut path or relative path based on `work_dir`.
+- `path` MUST be Unix style path but SHOULD NOT contain the prefix `/` for object storage services.

```

Comment:
When users call `store.Read`, they don't know whether this storage is object storage or not. So I think it's not a good rule.

And it conflicts with the example in the following.